datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -1056,6 +1069,81 @@ private void onDatastreamChange(List<DatastreamGroup> datastreamGroups) {
+        Connector connectorInstance = _connectors.get(connectorType).getConnector().getConnectorInstance();
+
+        Map<String, Optional<DatastreamGroupPartitionsMetadata>> datastreamPartitions =
+            connectorInstance.getDatastreamPartitions();
+
+        List<String> toMoveDatastream = _adapter.getDatastreamsNeedPartitionMovement(connectorType);
+
+        // retrieve the live datastreamGroups for validation
+        List<DatastreamGroup> toProcessedDatastreamGroups =
+            fetchDatastreamGroups().stream().filter(group1 -> connectorType.equals(group1.getConnectorName()))
+                .filter(group2 -> toMoveDatastream.contains(group2.getName()))
+                .filter(group3 -> datastreamPartitions.keySet().contains(group3.getName()))
+                .collect(Collectors.toList());
+
+        for (DatastreamGroup dg : toProcessedDatastreamGroups) {
+          DatastreamGroupPartitionsMetadata subscribedPartitions = connectorInstance.getDatastreamPartitions().get(dg.getName())
+              .orElseThrow(() -> new DatastreamTransientException("partition listener is not ready yet for datastream " + dg.getName()));
+          Map<String, Set<String>> suggestedAssignment = _adapter.getPartitionMovement(dg.getConnectorName(), dg.getName());
+          assignmentByInstance = strategy.movePartitions(assignmentByInstance, suggestedAssignment,
+              subscribedPartitions);
+          toCleanup.add(dg);
+        }
+      }
+
+      for (String key : assignmentByInstance.keySet()) {
+        newAssignmentsByInstance.put(key, new ArrayList<>(assignmentByInstance.get(key)));
+      }
+
+      _adapter.updateAllAssignments(newAssignmentsByInstance);
+
+      //clean up stored target assignment after the assignment is updated
+      for (DatastreamGroup dg : toCleanup) {
+        _adapter.cleanUpPartitionMovement(dg.getConnectorName(), dg.getName());
+      }
+
+      _log.info("Partition movement completed: datastreamGroup, assignment {} ", assignmentByInstance);
+
+      shouldRetry = false;
+    } catch (DatastreamTransientException ex) {
+      _log.info("Partition movement failed, retry again after a configurable period", ex);
+      shouldRetry = true;
+    } catch (Exception ex) {
+      //We do not retry if it is not transient exception
+      _log.info("Partition movement failed, Exception: ", ex);
+    }
+    if (!shouldRetry) {
+      _adapter.cleanupOldUnusedTasks(previousAssignmentByInstance, newAssignmentsByInstance);
+      updateCounterForMaxPartitionInTask(newAssignmentsByInstance);
+      _dynamicMetricsManager.createOrUpdateMeter(MODULE, NUM_PARTITION_MOVEMENTS, 1);
+    }  else if (!leaderPartitionMovementScheduled.get()) {
+      _log.info("Schedule retry for leader movement tasks");
+      _dynamicMetricsManager.createOrUpdateMeter(MODULE, "handleLeaderPartitionMovement", NUM_RETRIES, 1);
+      leaderPartitionMovementScheduled.set(true);
```

Comment:
The flag is to avoid to schedule multiple retries. This is atomicInteger so I don't think we need to make it volatile? Also, this variable should be only access in the coordinator (single threaded manner)